src/ORM/DataQuery.php

```diff
@@ -915,6 +917,19 @@ protected function joinHasManyRelation(
             return;
         }
 
+        if (class_exists(Versioned::class)) {
+            /** @var Versioned $versioned */
+            $versioned = Injector::inst()->get(Versioned::class);
+            $stage = $versioned->get_stage();
+        } else {
+            $versioned = null;
+            $stage = null;
+        }
```

Comment:
I'd really prefer to see this in an extension in the versioned module. If we do keep it here, I think there's some code duplication in this file that could be mitigated with a private method to apply these rules.